src/vos/vos_csum.c

```diff
@@ -0,0 +1,398 @@
+			 bsgl->bs_iovs[add_start + add_offset].bi_data_len);
+		sgl->sg_iovs[sgl_idx].iov_buf_len =
+			bsgl->bs_iovs[add_start + add_offset].bi_data_len;
+		sgl->sg_iovs[sgl_idx].iov_len =
+			bsgl->bs_iovs[add_start + add_offset].bi_data_len;
+		*add_idx += bsgl->bs_iovs[add_start + add_offset].bi_data_len;
+		add_offset++;
+		sgl->sg_nr++;
+	}
+	return add_offset;
+}
+
+static unsigned int
+calc_csum_params(struct dcs_csum_info *csum_info, struct csum_recalc *recalc,
+		 unsigned int prefix_len, unsigned int suffix_len,
+		 unsigned int rec_size)
+{
+
+	unsigned int cs_cnt, low_idx = recalc->cr_log_ext.ex_lo -
+							prefix_len / rec_size;
+	unsigned int high_idx = recalc->cr_log_ext.ex_hi +
+							suffix_len / rec_size;
+
+	assert(prefix_len % rec_size == 0);
+	cs_cnt = csum_chunk_count(recalc->cr_phy_ent->pe_csum_info.cs_chunksize,
+				    low_idx, high_idx, rec_size);
+	csum_info->cs_nr = cs_cnt;
+	D_ASSERT(cs_cnt * csum_info->cs_len  <= csum_info->cs_buf_len);
+	return low_idx;
+}
+
+static bool
+csum_agg_verify(struct csum_recalc *recalc, struct dcs_csum_info *new_csum,
+		unsigned int rec_size, unsigned int prefix_len)
+{
+	unsigned int j = 0;
+	bool match;
+
+	if (new_csum->cs_nr != recalc->cr_phy_ent->pe_csum_info.cs_nr) {
+		unsigned int chunksize = new_csum->cs_chunksize;
+		unsigned int orig_offset =
+			(recalc->cr_phy_ent->pe_rect.rc_ex.ex_lo +
+			 recalc->cr_phy_ent->pe_off)  * rec_size;
+		unsigned int out_offset = recalc->cr_log_ext.ex_lo * rec_size -
+								prefix_len;
+
+		D_ASSERT(new_csum->cs_nr <
+				 recalc->cr_phy_ent->pe_csum_info.cs_nr);
+		D_ASSERT(orig_offset <= out_offset);
+		if (orig_offset != out_offset) {
+			unsigned int add_start = chunksize -
+							orig_offset % chunksize;
+			unsigned int offset = orig_offset + add_start;
+
+			if (add_start)
+				j++;
+			while (offset < out_offset) {
+				offset += chunksize;
+				j++;
+			}
+			D_ASSERT(offset == out_offset);
+		}
+	}
+
+	match = memcmp(new_csum->cs_csum,
+		&recalc->cr_phy_ent->pe_csum_info.cs_csum[j * new_csum->cs_len],
+			new_csum->cs_nr * new_csum->cs_len) == 0;
+	if (!match)
+		return false;
+	return true;
+}
+
+static void
+csum_agg_recalc(void *recalc_args)
+{
+	d_sg_list_t		 sgl;
+	struct csum_recalc_args *args = recalc_args;
+	struct bio_sglist	*bsgl = args->cra_bsgl;
+	struct evt_entry_in	*ent_in = args->cra_ent_in;
+	struct csum_recalc	*recalcs = args->cra_recalcs;
+	struct daos_csummer	*csummer;
+	struct dcs_csum_info	 csum_info = args->cra_ent_in->ei_csum;
+	unsigned int		 buf_idx = 0;
+	unsigned int		 add_idx = 0;
+	unsigned int		 i,  add_offset = 0;
+	int			 rc = 0;
+
+	/* need at most prefix + buf + suffix in sgl */
+	D_ALLOC_ARRAY(sgl.sg_iovs, 3);
+	if (sgl.sg_iovs == NULL) {
+		args->cra_rc = -DER_NOMEM;
+		return;
+	}
+	daos_csummer_type_init(&csummer, csum_info.cs_type,
+			       csum_info.cs_chunksize);
+	for (i = 0; i < args->cra_seg_cnt; i++) {
+		bool		is_valid = false;
+		unsigned int	this_buf_nr, this_buf_idx;
+
+
```

Comment:
Again, please add comments to this function to explain each step.

Reply:
Comments have been added.